pageserver/src/tenant_mgr.rs

```diff
@@ -89,18 +122,26 @@ pub fn register_relish_download(
         tenant_id,
         timeline_id
     );
-    match access_repository().entry(tenant_id) {
-        Entry::Occupied(o) => init_timeline(o.get().as_ref(), timeline_id),
-        Entry::Vacant(v) => {
-            log::info!("New repo initialized");
-            let new_repo = init_repo(conf, tenant_id);
-            init_timeline(new_repo.as_ref(), timeline_id);
-            v.insert(new_repo);
+
+    {
+        let mut m = access_tenants();
+        let mut tenant = m.get_mut(&tenant_id).unwrap();
+        tenant.state = TenantState::Downloading;
+        match &tenant.repo {
+            Some(repo) => init_timeline(repo, timeline_id),
+            None => {
+                log::info!("Initialize new repo");
+            }
         }
     }
+
+    // init repo updates Tenant state
+    init_repo(conf, tenant_id);
+    let new_repo = get_repository_for_tenant(tenant_id).unwrap();
+    init_timeline(&new_repo, timeline_id);
 }
 
-fn init_timeline(repo: &dyn Repository, timeline_id: ZTimelineId) {
+fn init_timeline(repo: &Arc<dyn Repository>, timeline_id: ZTimelineId) {
```

Comment:
NIT: `&Arc<dyn Repository>` looks odd, I think it should be the previous one still, `repo: &dyn Repository` and use https://doc.rust-lang.org/std/sync/struct.Arc.html#method.as_ref to properly pass it.